perses/protocols/nonequilibrium_cycling.py

```diff
@@ -0,0 +1,452 @@
+        cumulated_reverse_work = reverse_work[-1] - reverse_work[0]
+
+        return {"forward_work": cumulated_forward_work,
+                "reverse_work": cumulated_reverse_work,
+                "paths": {"forward_work": simulations[0].outputs['forward_work'],
+                          "reverse_work": simulations[0].outputs['reverse_work']},
+                }
+
+
+class NonEquilibriumCyclingProtocolResult(ProtocolResult):
+    """
+
+    """
+
+    def get_estimate(self, n_bootstraps=1000):
+        """
+        Get a free energy estimate using bootstrap and BAR.
+
+        Parameters
+        ----------
+        n_bootstraps: int
+            Number of bootstrapped samples to use.
+
+        Returns
+        -------
+        free_energy: float
+            Free energy estimate in units of kT.
+
+        """
+        # free_energy, error = pymbar.bar.BAR(forward_work, reverse_work)
+        # self.data  # this has the returned object from _gather
+        import numpy as np
+
+        forward: List[float] = self.data["forward_work"]
+        reverse: List[float] = self.data["reverse_work"]
+
+        all_dgs = self._do_bootstrap(forward, reverse, n_bootstraps=n_bootstraps)
+
+        return np.mean(all_dgs)
+
+    def get_uncertainty(self, n_bootstraps=1000):
+        """
+        Estimate uncertainty using standard deviation of the distribution of bootstrapped
+        free energy (dg) samples.
+
+        Parameters
+        ----------
+        n_bootstraps
+
+        Returns
+        -------
+        free_energy_uncertainty: float
+            Uncertainty on the free energy estimate in units of kT.
+
+        """
+        import numpy as np
+
+        forward: List[float] = self.data["forward_work"]
+        reverse: List[float] = self.data["reverse_work"]
+
+        all_dgs = self._do_bootstrap(forward, reverse, n_bootstraps=n_bootstraps)
+
+        # TODO: Check if standard deviation is a good uncertainty estimator
+        return np.std(all_dgs)
+
+    def get_rate_of_convergence(self):
+        ...
+
+    @lru_cache()
+    def _do_bootstrap(self, forward, reverse, n_bootstraps=1000):
+        """
+        Performs bootstrapping from forward and reverse cumulated works.
+
+        Returns
+        -------
+        free_energies: List[Float]
+            List of bootstrapped free energies in units of kT.
+        """
+        import pymbar
+        import numpy as np
+
+        all_dgs = []
+        # all_ddgs = []
+
+        for i in range(n_bootstraps):
+            subsample_indices_forward = np.random.choice(len(forward), forward)
+            subsample_indices_reverse = np.random.choice(len(reverse), reverse)
+            forward_chosen = forward[subsample_indices_forward]
+            reverse_chosen = reverse[subsample_indices_reverse]
+            dg, ddg = pymbar.bar.BAR(forward_chosen, reverse_chosen)
+            all_dgs.append(dg)
+            # all_ddgs.append(ddg)
+
+        return all_dgs
```

Comment:
@jchodera We need your input here. Especially, when computing the free energy uncertainty we are doing it via the computing the standard deviation of the bootstrapped DeltaG distribution. Does that seem most appropriate?

Reply:
For nonequilibrium cycling, the forward and reverse work values are paired since they come from the same trajectory. Each trajectory can be assumed to be the unit to bootstrap on. SO I'd change this to:
```python
        # Check to make sure forward and reverse work values match in length
        assert len(forward) == len(reverse), "Forward and reverse work values are not paired"
        ntraj = len(forward)
        for i in range(n_bootstraps):
            # Sample trajectory indices with replacement
            indices = np.random.choice(np.arange(ntraj), size=[ntraj], replace=True)
            # Compute and store BAR estimates
            dg, ddg = pymbar.bar.BAR(forward[indices], reverse[indices])
            all_dgs.append(dg)
```
Also, is it critical to return a list here, rather than a numpy array? Why not return a numpy array, especially since you know the number of bootstrap samples you will generate anyway.

Reply:
For `get_estimate`, there is no need to bootstrap---you should just return the MLE. Please replace with this:
```python
    def get_estimate(self):
        """
        Get a free energy estimate.

        Returns
        -------
        free_energy: float
            Free energy estimate in units of kT.
        """
        free_energy, error = pymbar.bar.BAR(forward_work, reverse_work)
        return free_energy
```

Reply:
Also, you don't want to bootstrap every time you call `get_uncertainty`. I'd cache that result after you compute it the first time, since it's expensive.

I'm also not quite sure why we split up the accessors for free energy and uncertainty. Wouldn't we want to compute them both at the same time, once, and then cache them?

Reply:
> Also, you don't want to bootstrap every time you call `get_uncertainty`. I'd cache that result after you compute it the first time, since it's expensive. You may also want to cache the results of `get_free_energy` since BAR incurs some computation time.

In this case we are using a cache via the `@lru_cache()` decorator in https://github.com/choderalab/perses/pull/1066/files#diff-2d23d1987b5a674a614a5af699d2b64c5befb5c1c2a0b487200749cf5569b042R360

Reply:
Great!

Reply:
Thanks so much for the improvements suggested. They have been addressed in commit https://github.com/choderalab/perses/pull/1066/commits/c17c7eb8c06bf4242cbc52b4a27ee8ccb4f417a6